src/Rector/Architecture/RepositoryAsService/ServiceLocatorToDIRector.php

```diff
@@ -0,0 +1,129 @@
+use Rector\NodeAnalyzer\MethodCallAnalyzer;
+use Rector\Rector\AbstractRector;
+
+final class ServiceLocatorToDIRector extends AbstractRector
+{
+    /**
+     * @var MethodCallAnalyzer
+     */
+    private $methodCallAnalyzer;
+
+    /**
+     * @var PropertyFetchNodeFactory
+     */
+    private $propertyFetchNodeFactory;
+
+    /**
+     * @var RepositoryForDoctrineEntityProviderInterface
+     */
+    private $repositoryForDoctrineEntityProvider;
+
+    /**
+     * @var ClassPropertyCollector
+     */
+    private $classPropertyCollector;
+
+    public function __construct(
+        MethodCallAnalyzer $methodCallAnalyzer,
+        PropertyFetchNodeFactory $propertyFetchNodeFactory,
+        RepositoryForDoctrineEntityProviderInterface $repositoryForDoctrineEntityProvider,
+        ClassPropertyCollector $classPropertyCollector
+    ) {
+        $this->methodCallAnalyzer = $methodCallAnalyzer;
+        $this->propertyFetchNodeFactory = $propertyFetchNodeFactory;
+        $this->repositoryForDoctrineEntityProvider = $repositoryForDoctrineEntityProvider;
+        $this->classPropertyCollector = $classPropertyCollector;
+    }
+
+    public function isCandidate(Node $node): bool
+    {
+        if (! $this->methodCallAnalyzer->isMethod($node, 'getRepository')) {
+            return false;
+        }
+
+        $className = $node->getAttribute(Attribute::CLASS_NAME);
+
+        if ($className === null) {
+            return false;
+        }
+
+        return ! Strings::endsWith($node->getAttribute(Attribute::CLASS_NAME), 'Repository');
+    }
+
+    public function refactor(Node $node): ?Node
+    {
+        $this->classPropertyCollector->addPropertyForClass(
+            (string) $node->getAttribute(Attribute::CLASS_NAME),
+            [$this->repositoryFqn($node)],
+            $this->repositoryVariableName($node)
+        );
+
+        return $this->propertyFetchNodeFactory->createLocalWithPropertyName(
+            $this->repositoryVariableName($node)
+        );
+    }
+
+    private function repositoryVariableName(Node $node): string
+    {
+        return lcfirst($this->repositoryShortName($node));
+    }
+
+    private function repositoryShortName(Node $node): string
+    {
+        $nameSpaceParts = explode('\\', $this->repositoryFqn($node));
+        return end($nameSpaceParts);
+    }
+
+    private function repositoryFqn(Node $node): string
+    {
+        $entityFqnOrAlias = $this->entityFqnOrAlias($node);
+
+        $repositoryClassName = $this->repositoryForDoctrineEntityProvider->provideRepositoryForEntity(
+            $entityFqnOrAlias
+        );
+
+        if ($repositoryClassName === null) {
+            throw new RectorProviderException(sprintf(
+                'A repository was not provided for "%s" entity by your "%s" class.',
+                $entityFqnOrAlias,
+                get_class($this->repositoryForDoctrineEntityProvider)
+            ));
+        }
+
+        return $repositoryClassName;
```

Comment:
As for early return, it's better to return simpler results first.
It's easier to read and also in one with code flow

```php
if ($repositoryClassName) {
   // all good = simple
   return $repositoryClassName;
}

// oh no, messy part
throw ...
```